// esniff/test/__playground/function.js
require('on\u0065')
require(12)
var foo = require('thr/ee').b.c.d().e.f().g
if (foobar) {
	var bla = function() {
		switch (baz) {
		default:
			require("fo\\ur")()()()()
		}
	}
	qux(1, 3, quux(require('five').bar.foo, 4), 2)
}

function func() {
	return
}

require(baz);

var str = "raz; require('string'); ";

var str2 = 'raz; require("string2");';
var a = b
	/require("six")/g;

a['raz']
	/require("seven")/g

var c = d;
	/require('regexp');/g

require()

if (a) {
}
	/require("reqexp2")/

var x = {
	raz: 'dwa',
	marko: { tile: require('nested-one') },
	biurko: { elos: require('nested-one2') }
}

/** comment **/

obj.require('object');

obj.
	require("object2");

// require('comment');
var raz = 4;  // require('comment2')

/* require('mlcomment'); */

function test() {
	return require('nine');
}

/* comment

 require('mlcomment2');

 */

if (foo) { } require('ten');

otheRrequire('marko');

require('eleven' + 'split' +
	'path');

require(true ? "twelve" : "bar");

require("object3" + { foo: bar() });

require.foo;


function test() {
	a();return    require(('four') + 'teen');
}


function test() {
	return require/* raz */('fifteen');
}

(require)('donottake')

function test() {
	a(); return (require('sixteen'));
}

switch (require('seventeen')) {
case 1:
	require("'eighteen'");
	break;
case 2:
	require('nineteen');
	break;
default:
	require('twenty');
}

if (require('twenty/one')) {
} else if (require('twenty/two')) {
	i = (require('twenty/three'));
}

for (var i, j = require('/twenty/two/2/'); require('twenty/three/2/');
	require('twenty/four/2/\'')) {
	require("twenty/five/2/\"");
}

for (i in require('\'twenty/seven\'')) {
	require("\"twenty/eight");
}

with (a, require("\"twenty/nine\"")) {
	require('"thirty"');
}
isNaN();require('mid-thirty')

require('inner' + require(require('marko')) + 'elo')
require('thirty\
	break-line \
	one');

module.exports = /"/, require('thirty\two')
