// traceur/dist/commonjs/util-imports.js
"use strict";
var $__util_47_MutedErrorReporter_46_js__,
    $__WebPageTranscoder_46_js__,
    $__HTMLImportTranscoder_46_js__,
    $__Options_46_js__,
    $__util_47_ErrorReporter_46_js__,
    $__util_47_CollectingErrorReporter_46_js__;
($__util_47_MutedErrorReporter_46_js__ = require("./util/MutedErrorReporter.js"), $__util_47_MutedErrorReporter_46_js__ && $__util_47_MutedErrorReporter_46_js__.__esModule && $__util_47_MutedErrorReporter_46_js__ || {default: $__util_47_MutedErrorReporter_46_js__});
var $__WebPageTranscoder_46_js__ = ($__WebPageTranscoder_46_js__ = require("./WebPageTranscoder.js"), $__WebPageTranscoder_46_js__ && $__WebPageTranscoder_46_js__.__esModule && $__WebPageTranscoder_46_js__ || {default: $__WebPageTranscoder_46_js__});
var $__HTMLImportTranscoder_46_js__ = ($__HTMLImportTranscoder_46_js__ = require("./HTMLImportTranscoder.js"), $__HTMLImportTranscoder_46_js__ && $__HTMLImportTranscoder_46_js__.__esModule && $__HTMLImportTranscoder_46_js__ || {default: $__HTMLImportTranscoder_46_js__});
var $__0 = ($__Options_46_js__ = require("./Options.js"), $__Options_46_js__ && $__Options_46_js__.__esModule && $__Options_46_js__ || {default: $__Options_46_js__}),
    addOptions = $__0.addOptions,
    CommandOptions = $__0.CommandOptions,
    Options = $__0.Options;
var ErrorReporter = ($__util_47_ErrorReporter_46_js__ = require("./util/ErrorReporter.js"), $__util_47_ErrorReporter_46_js__ && $__util_47_ErrorReporter_46_js__.__esModule && $__util_47_ErrorReporter_46_js__ || {default: $__util_47_ErrorReporter_46_js__}).ErrorReporter;
var CollectingErrorReporter = ($__util_47_CollectingErrorReporter_46_js__ = require("./util/CollectingErrorReporter.js"), $__util_47_CollectingErrorReporter_46_js__ && $__util_47_CollectingErrorReporter_46_js__.__esModule && $__util_47_CollectingErrorReporter_46_js__ || {default: $__util_47_CollectingErrorReporter_46_js__}).CollectingErrorReporter;
var util = {
  addOptions: addOptions,
  CommandOptions: CommandOptions,
  CollectingErrorReporter: CollectingErrorReporter,
  ErrorReporter: ErrorReporter,
  Options: Options
};
Object.defineProperties(module.exports, {
  WebPageTranscoder: {get: function() {
      return $__WebPageTranscoder_46_js__.WebPageTranscoder;
    }},
  HTMLImportTranscoder: {get: function() {
      return $__HTMLImportTranscoder_46_js__.HTMLImportTranscoder;
    }},
  util: {get: function() {
      return util;
    }},
  __esModule: {value: true}
});
